Gems/AtomTressFX/Assets/Shaders/HairSimulationCommon.azsli

```diff
@@ -0,0 +1,446 @@
+
+    if ( lengthSqr < 0.001 )
+        return float4(0, 0, 0, 1.0f);
+
+    q.x = -q.x / lengthSqr;
+    q.y = -q.y / lengthSqr;
+    q.z = -q.z / lengthSqr;
+    q.w = q.w / lengthSqr;
+
+    return q;
+}
+
+float3 MultQuaternionAndVector(float4 q, float3 v)
+{
+    float3 uv, uuv;
+    float3 qvec = float3(q.x, q.y, q.z);
+    uv = cross(qvec, v);
+    uuv = cross(qvec, uv);
+    uv *= (2.0f * q.w);
+    uuv *= 2.0f;
+
+    return v + uv + uuv;
+}
+
+float4 MultQuaternionAndQuaternion(float4 qA, float4 qB)
+{
+    float4 q;
+
+    q.w = qA.w * qB.w - qA.x * qB.x - qA.y * qB.y - qA.z * qB.z;
+    q.x = qA.w * qB.x + qA.x * qB.w + qA.y * qB.z - qA.z * qB.y;
+    q.y = qA.w * qB.y + qA.y * qB.w + qA.z * qB.x - qA.x * qB.z;
+    q.z = qA.w * qB.z + qA.z * qB.w + qA.x * qB.y - qA.y * qB.x;
+
+    return q;
+}
+
+float4 NormalizeQuaternion(float4 q)
+{
+    float4 qq = q;
+    float n = q.x*q.x + q.y*q.y + q.z*q.z + q.w*q.w;
+
+    if (n < 1e-10f)
+    {
+        qq.w = 1;
+        return qq;
+    }
+
+    qq *= 1.0f / sqrt(n);
+    return qq;
+}
+
+// Compute a quaternion which rotates u to v. u and v must be unit vector. 
+float4 QuatFromTwoUnitVectors(float3 u, float3 v)
+{
+    float r = 1.f + dot(u, v);
+    float3 n;
+
+    // if u and v are parallel
+    if (r < 1e-7)
+    {
+        r = 0.0f;
+        n = abs(u.x) > abs(u.z) ? float3(-u.y, u.x, 0.f) : float3(0.f, -u.z, u.y);
+    }
+    else
+    {
+        n = cross(u, v);  
+    }
+
+    float4 q = float4(n.x, n.y, n.z, r);
+    return NormalizeQuaternion(q);
+}
+
+// Make the inpute 4x4 matrix be identity
+float4x4 MakeIdentity()
+{
+    float4x4 m;
+    m._m00 = 1;   m._m01 = 0;   m._m02 = 0;   m._m03 = 0;
+    m._m10 = 0;   m._m11 = 1;   m._m12 = 0;   m._m13 = 0;
+    m._m20 = 0;   m._m21 = 0;   m._m22 = 1;   m._m23 = 0;
+    m._m30 = 0;   m._m31 = 0;   m._m32 = 0;   m._m33 = 1;
+
+    return m;
+}
+
+void ApplyDistanceConstraint(inout float4 pos0, inout float4 pos1, float targetDistance, float stiffness = 1.0)
+{
+    float3 delta = pos1.xyz - pos0.xyz;
+    float distance = max(length(delta), 1e-7);
+    float stretching = 1 - targetDistance / distance;
+    delta = stretching * delta;
+    float2 multiplier = ConstraintMultiplier(pos0, pos1);
+
+    pos0.xyz += multiplier[0] * delta * stiffness;
+    pos1.xyz -= multiplier[1] * delta * stiffness;
+}
+
+void CalcIndicesInVertexLevelTotal(uint local_id, uint group_id, inout uint globalStrandIndex, inout uint localStrandIndex, inout uint globalVertexIndex, inout uint localVertexIndex, inout uint numVerticesInTheStrand, inout uint indexForSharedMem, inout uint strandType)
+{
+    indexForSharedMem = local_id;
+    numVerticesInTheStrand = (THREAD_GROUP_SIZE / g_NumOfStrandsPerThreadGroup);
+
+    localStrandIndex = local_id % g_NumOfStrandsPerThreadGroup;
+    globalStrandIndex = group_id * g_NumOfStrandsPerThreadGroup + localStrandIndex;
+    localVertexIndex = (local_id - localStrandIndex) / g_NumOfStrandsPerThreadGroup;
```

Comment:
This indexing is part of the original TressFX code and is done this way (originally) due to the way the data is being exported and stored. 
Since this convention is used everywhere following the data schema, changing this would take quite some time and will be very shader error prone, hence at this point in time will have very low priority.